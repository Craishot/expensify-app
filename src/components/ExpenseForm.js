import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

// ExpenseFrom component for use in user adding and editing expenses
export default class ExpenseForm extends React.Component {
    constructor(props) {
        // Dont break stuffs
        super(props);

        // Setup component initial state
        this.state = {
            description: props.expense ? props.expense.description : '',
            note: props.expense ? props.expense.note : '',
            amount: props.expense ? (props.expense.amount / 100).toString() : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calendarFocused: false,
            error: ''
        }
    }

    // Handler for change in description text input 
    onDescriptionChange = (e) => {
        // Get new description value from event object and save it to the component state
        const description = e.target.value;
        this.setState(() => ({ description }));
    };

    // Handler for change in amount number input
    onAmountChange = (e) => {
        const amount = e.target.value;

        if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount }));
        }
    };

    // Handler for change in note textarea
    onNoteChange = (e) => {
        // Get new note value from event object and save it to the component state
        const note = e.target.value;
        this.setState(() => ({ note }));
    };

    // Handler for date change for the SingleDatePicker component
    onDateChange = (createdAt) => {
        if(createdAt) {
            this.setState(() => ({ createdAt }))
        }
    };

    // Handler for change in focus for the SingleDatePicker component
    onFocusChange = ({ focused }) => {
        this.setState(() => ({ calendarFocused: focused }));
    };

    // Handler for form submit
    onSubmit = (e) => {
        // Prevent default behavior on form submit
        e.preventDefault();

        //
        if(!this.state.description || !this.state.amount) {
            // Update error state to display error message to the user
            this.setState(() => ({ error: "Please provide description and amount." }));
        }
        else {
            // Clear error state and submit form with correct data
            this.setState(() => ({ error: '' }));

            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100, 
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            });
        }
    };

    // Component render function
    render() {
        return (
            <div>
                {}
                { this.state.error && <p>{ this.state.error }</p> }
                <form onSubmit={ this.onSubmit }>
                    { /* Expense Description Text Input */ }
                    <input 
                        type="text"
                        placeholder="Description"
                        value={ this.state.description }
                        onChange={ this.onDescriptionChange }
                        autoFocus
                    />
                    { /* Expense Amount Number Input */ }
                    <input 
                        type="text"
                        placeholder="Amount"
                        value={ this.state.amount }
                        onChange={ this.onAmountChange }
                    />
                    { /* Expense Date Picker (third party component) */ }
                    <SingleDatePicker 
                        date={ this.state.createdAt }
                        onDateChange={ this.onDateChange }
                        focused={ this.state.calendarFocused }
                        onFocusChange={ this.onFocusChange }
                        numberOfMonths={ 1 }
                        isOutsideRange={ () => false }
                    />
                    { /* Expense Note Text Area */ }
                    <textarea
                        placeholder="Add a note for your expense (optional)"
                        onChange={ this.onNoteChange }
                    >
                    </textarea>
                    <button>Add Expense</button>
                </form>
            </div>
        );
    }
}