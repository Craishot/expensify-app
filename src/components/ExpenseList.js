import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

// Expense List React component
export const ExpenseList = (props) => (
    <div>
        <h1>Expense List</h1>
        {
            props.expenses.length === 0 ? (
                <p>No expenses</p>
            ) : (
                props.expenses.map((expense) => {
                    return <ExpenseListItem key={ expense.id } { ...expense } />
                })
            )
        }
    </div>
);

// Map component state to the redux store and export it as the default
const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    };
};

export default connect(mapStateToProps)(ExpenseList);