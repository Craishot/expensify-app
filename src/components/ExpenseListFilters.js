import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import { setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate } from '../actions/filters'; 

// Expense List Filters Component
export class ExpenseListFilters extends React.Component {
    state = {
        calendarFocused: null
    };

    // Handle date change from the date range picker component
    onDatesChange = ({ startDate, endDate }) => {
        // Dispatch actions to change the start and end dates for the filter in the redux store
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    };

    // Handle focus change for the date range picker component
    onFocusChange = (calendarFocused) => {
        this.setState(() => ({ calendarFocused }));
    };

    // Handle change in the text filter input
    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value);
    };

    //
    onSortChange = (e) => {
        if(e.target.value === 'amount') {
            this.props.sortByAmount();
        } 
        else if(e.target.value === 'date') {
            this.props.sortByDate();
        }
        else {
            throw 'Invalid sortBy type...'
        }
    };

    render() {
        return (
            <div>
                { /* Text filter */ }
                <input 
                    type='text' 
                    value={ this.props.filters.text } 
                    onChange={ this.onTextChange }
                />
                { /* Sort by dropdown */ }
                <select 
                    value={ this.props.filters.sortBy } 
                    onChange={ this.onSortChange }>
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>
                </select>
                { /* Date range filter */ }
                <DateRangePicker 
                    startDate={ this.props.filters.startDate }
                    endDate={ this.props.filters.endDate }
                    onDatesChange={ this.onDatesChange }
                    focusedInput={ this.state.calendarFocused }
                    onFocusChange={ this.onFocusChange }
                    showClearDates={ true }
                    numberOfMonths={ 1 }
                    isOutsideRange={ () => false }
                />
            </div>
        );
    };
};

// Give Expense List Filters component access to redux store
const mapStateToProps = (state) => {
    return {
        filters: state.filters
    };
};

// Map dispatch function to the props
const mapDispatchToProps = (dispatch) => {
    return {
        setTextFilter: (text) => dispatch(setTextFilter(text)),
        sortByDate: () => dispatch(sortByDate()),
        sortByAmount: () => dispatch(sortByAmount()),
        setStartDate: (startDate) => dispatch(setStartDate(startDate)),
        setEndDate: (endDate) => dispatch(setEndDate(endDate))
    };
};


// Default export
export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);