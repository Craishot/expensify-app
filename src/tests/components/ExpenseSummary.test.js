import React from 'react';
import { shallow } from 'enzyme';

import { ExpenseSummary } from '../../components/ExpenseSummary';

test('should correctly render ExpenseSummary with 1 expense', () => {
    const wrapper = shallow(<ExpenseSummary expenseCount={ 1 } expenseTotal={ 50000 } />);
    expect(wrapper).toMatchSnapshot();
});

test('should correctly render ExpenseSummary with multiple expenses', () => {
    const wrapper = shallow(<ExpenseSummary expenseCount={ 24 } expenseTotal={ 1052468752 } />);
    expect(wrapper).toMatchSnapshot();
});