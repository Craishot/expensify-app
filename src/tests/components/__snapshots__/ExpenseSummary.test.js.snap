// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correctly render ExpenseSummary with 1 expense 1`] = `
<div>
  <h1>
    Viewing 
    1
     
    expense
     totalling 
    $500.00
  </h1>
</div>
`;

exports[`should correctly render ExpenseSummary with multiple expenses 1`] = `
<div>
  <h1>
    Viewing 
    24
     
    expenses
     totalling 
    $10,524,687.52
  </h1>
</div>
`;
